<a href="https://www.kaggle.com/elakapoor/descriptive-statistics-using-pandas?scriptVersionId=87973856" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

<h1><center>Analysis of heart attacks</center></h1>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import plotly.express as px
from scipy.stats import skew, kurtosis
import plotly.figure_factory as ff

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/opt/conda/lib/python3.7/site-packages/geopandas/_compat.py:115: UserWarning: The Shapely GEOS version (3.9.1-CAPI-1.14.2) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
  shapely_geos_version, geos_capi_version_string


/kaggle/input/heart-attack-analysis-prediction-dataset/o2Saturation.csv
/kaggle/input/heart-attack-analysis-prediction-dataset/heart.csv



1. **Age** : Age of the patient
2. **Sex** : Sex of the patient
3. **exang**: exercise induced angina (1 = yes; 0 = no)
4. ca: number of major vessels (0-3)
5. **cp** : Chest Pain type chest pain type
    6. Value 1: typical angina
    7. Value 2: atypical angina
    8. Value 3: non-anginal pain
    9. Value 4: asymptomatic
10. **trtbps** : resting blood pressure (in mm Hg)
11. **chol** : cholestoral in mg/dl fetched via BMI sensor
12. **fbs** : (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
13. **rest_ecg** : resting electrocardiographic results
    14. Value 0: normal
    15. Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
    16. Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
17. thalach : maximum heart rate achieved
18. target : 0= less chance of heart attack 1= more chance of heart attack



In [2]:
path = "/kaggle/input/heart-attack-analysis-prediction-dataset/heart.csv"
df = pd.read_csv(path)
df = df[["age", "sex", 'exng', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'output']]
df.head()

age  sex  exng  cp  trtbps  chol  fbs  restecg  output
0   63    1     0   3     145   233    1        0       1
1   37    1     0   2     130   250    0        1       1
2   41    0     0   1     130   204    0        0       1
3   56    1     0   1     120   236    0        1       1
4   57    0     1   0     120   354    0        1       1

In [3]:
df.columns

Index(['age', 'sex', 'exng', 'cp', 'trtbps', 'chol', 'fbs', 'restecg',
       'output'],
      dtype='object')

In [4]:
len(df)

303

# Central Tendency
It is measured using 3M's that is mean, median and mode.
1. Mean: It is defined as the average of the values present.
2. Median: It is the centrally located value of the dataset when arranged in ascending order.
3. Mode: It is the most frequent value in the dataset.<br>
Now the question arises how are these values useful in data analysis of data science. To answer it let us see the following graph.


In [5]:
x = df["age"]
hist_data = [x]
group_labels = ['age'] # name of the dataset

fig = ff.create_distplot(hist_data, group_labels, show_rug=False)
fig.update_layout(title = "Distribution of age")
fig.show()

print("Mean age:", df["age"].mean())
print("Median age:", df["age"].median())
print("Mode of age:", df["age"].mode())
print("Skewness: ", skew(df["age"]))

Mean age: 54.366336633663366
Median age: 55.0
Mode of age: 0    58
dtype: int64
Skewness:  -0.2014596901125439


From the graph above we have following observations:
1. The average value is around 54 years
2. The centrally located vale is median which is 55
3. The most frequent values is 58<br>
Now we can see that the curve is not a perfect gaussian curve or bell shaped curve. The tail of the curve is towards left so it is a left skew curve (negative value above). So how does these values justify the curve shape? If we arrange our values for mean, median and mode we see that:<br>
Mode > Median > Mean (shows left skewness)<br>
Mode < Median < Mean (shows right skewness)<br>
Mode = Median = Mean (perfect bell shape curve)<br>
Hence these values provides the knowledge about the shape of the data distribution curve. Which makes it easier to deal with the data. 
Also, we see that mean is nearly equal to median which can tell us that there maybe no outliers present. Let us confirm it.

In [6]:
# boxplot represent presence or absence of outliers
fig = px.box(df, x = "age", title = "distribution of age")
fig.show()

# Spread
It is a variability of the data within a distribution.This spread is the distribution of the data around the central tendency. It can be measured using the following metrics:<br>
1. Range
2. Quartile
3. Variance
4. Standard Deviation<br>


**Range:** It can be measured as the difference between the largest and smallest values

In [7]:
range_chol = df["chol"].max() - df["chol"].min()
print("The range of the cholestrol level is:",range_chol)

The range of the cholestrol level is: 438


In [8]:
x = df["chol"]
hist_data = [x]
group_labels = ['cholestrol'] # name of the dataset

fig = ff.create_distplot(hist_data, group_labels, show_rug=False)
fig.update_layout(title = "Distribution of cholestrol levels")
fig.show()

Here the values are concentrated between 200-300 but we see the range is 438. The reason is because of outliers in the data.<br>
The curve is right skewed curve. Here as a practice you can calculate the central tendency and see the above equation for right skew matches.

**Quartiles:** As a name suggest it represent quarter of the data and divides the data in 
4 equal parts. Namely 25%(Q1), 50%(Q2), 75%(Q3). The data is arranged in ascending order
which is the reason median and Q2 are equal. <br>
**IQR** is a range between Q1 and Q3. IQR is preferred over a range as it is not influence by outliers. IQR is used to measure variability by splitting a data set into four equal quartiles.<br>
Find Outlier = [(Q1 – 1.5 * IQR), (Q3 + 1.5 * IQR)]<br>
Any number not in range is outlier<br>
The quartiles and outliers can be explained with the help of box plot.

In [9]:
fig = px.box(df, x = "chol", title = "Distribution of cholestrol levels")
fig.show()

In [10]:
df.chol.describe()

count    303.000000
mean     246.264026
std       51.830751
min      126.000000
25%      211.000000
50%      240.000000
75%      274.500000
max      564.000000
Name: chol, dtype: float64

In [11]:
# Outlier calculation
#IQR = Q3 - Q1
Q1 = 211.0
Q2 = 240.0
Q3 = 274.5
IQR = 274.5 - 211
outlier1 = (Q1 - 1.5 * IQR)
outlier2 = (Q3 + 1.5 * IQR)
print(f"The numbers outside the range of {outlier1} and {outlier2} will be considered as outliers")
print("The box plot verify our calculation. All the number greater than 369.75 are shown as outliers.")

The numbers outside the range of 115.75 and 369.75 will be considered as outliers
The box plot verify our calculation. All the number greater than 369.75 are shown as outliers.


**Variance:** It is a statistical parameter used to quantify spread. It measures how far each number in the set from the mean and thus from every other number in the set.
Observation near to mean value gets the lower result and far from means gets higher value.
1. A high variance indicates that the numbers are far from the mean and far from each other. 
2. A low variance indicates that the numbers are close to the mean and to each other. 
3. If variance is 0 that means that all the numbers in the dataset are the identical. 
4. The valid variance is always a positive number (0 or more).
<br>
**Standard Deviation:** It is a square root of variance. It is more commonly used because the unit measure is easy to calculate spread.<br>
For example in variance the unit is kg<sup>2</sup> whereas in standard deviation it is kg.

In [12]:
# case of very high variance as per the plot above
print("Variance: ",df["chol"].var())
print("Standard Deviation: ", df["chol"].std())

Variance:  2686.4267479728114
Standard Deviation:  51.830750987930045


# Normalization
In it values are shifted and rescaled so that they end up ranging between 0 and 1. It is also known as Min-Max scaling. When the curve does not follow the gaussian distribution it is good to do normalization.

In [13]:
df["normalized_chol"]=(df["chol"]-df["chol"].min())/(df["chol"].max()-df["chol"].min())

In [14]:
x = df["normalized_chol"]
hist_data = [x]
group_labels = ['normalized cholestrol'] # name of the dataset

fig = ff.create_distplot(hist_data, group_labels, show_rug=False)
fig.update_layout(title = "Distribution of normalized cholestrol levels")
fig.show()

Compared to the above curve (Fig:1) we conclude following:
1. The spread is less now
2. The height of the curve is increased
3. All values as in range of 0 to 1

# PMF

Probability mass function (PMF) maps each value to its corresponding probability. PMF is plotted for discrete distributions. <br> However, when there are too many values in the sample, it becomes difficult to read and interpret the PMF plot. Because with the decrease in the probability of each values, the random noise increases.

In [15]:
chol_pmf = df["chol"].value_counts(normalize = True)
chol_pmf

204    0.019802
197    0.019802
234    0.019802
269    0.016502
254    0.016502
         ...   
284    0.003300
224    0.003300
167    0.003300
276    0.003300
131    0.003300
Name: chol, Length: 152, dtype: float64

In [16]:
fig = px.bar(x=chol_pmf.index, y =chol_pmf.values)
fig.update_layout(yaxis_title = "Normalized cholestrol values",
                 xaxis_title = "Index number",
                 title = "PMF distribution of the cholestrol values")
fig.show()

The above notebook shows the various parametrs for the descriptive analysis of the data. The ideas for the notebook is taken from the book "Think Stats: Probability and
Statistics for Programmers by Allen B. Downey". While writing the notebook the help from various articles from Medium and Analytics Vidhya was also taken. <br>
I hope you enjoy this notebook. Please share your thoughts in comments and upvote if you find the work useful.